Release/src/uri/uri.cpp

```diff
@@ -423,6 +423,53 @@ namespace
         return encoded;
     }
 
+    // 5.2.3. Merge Paths https://tools.ietf.org/html/rfc3986#section-5.2.3
+    utility::string_t mergePaths(const utility::string_t &base, const utility::string_t &relative)
+    {
+        const auto lastSlash = base.rfind(_XPLATSTR('/'));
+        if (lastSlash == utility::string_t::npos)
+        {
+            return base + _XPLATSTR('/') + relative;
+        }
+        else if (lastSlash == base.size() - 1)
+        {
+            return base + relative;
+        }
+        // path contains and does not end with '/', we remove segment after last '/'
+        return base.substr(0, lastSlash + 1) + relative;
+    }
+
+    // 5.2.4. Remove Dot Segments https://tools.ietf.org/html/rfc3986#section-5.2.4
+    void removeDotSegments(web::uri_builder &builder)
+    {
+        if (builder.path().find(_XPLATSTR('.')) == utility::string_t::npos)
+            return;
+
+        const auto segments = web::uri::split_path(builder.path());
+        std::vector<std::reference_wrapper<const utility::string_t>> result;
+        for (auto& segment : segments)
+        {
+            if (segment == _XPLATSTR("."))
+                continue;
+            else if (segment != _XPLATSTR(".."))
+                result.push_back(segment);
+            else if (!result.empty())
+                result.pop_back();
+        }
+        if (result.empty())
+        {
+            builder.set_path(utility::string_t());
+            return;
+        }
+        utility::stringstream_t path;
```

Comment:
Ultimately either the `string` or the `stringstream` will manage the buffer growth so it is roughly the same.
I consider `stringstream`'s memory management is somewhat better because it is supposed to grow by its very purpose, and the interface is more convenient (e.g. you can `operator<<` numbers, though not in this case).
It's all arguable, however.

Reply:
I can tell you as someone who has maintained MSVC++'s stringstream that it is far less efficient than std::string. Like "virtual function calls per character" less efficient. If you need number formatting or something like that with it then of course stringstream is the only option, but this is just string concat.

Reply:
You are write, I forgot about that design feature of streams. Will fix it.